### ISSUE TITLE
CPU load misreported

### ISSUE THREAD

**Issue by sithglan** [vmware/govmomi#1878]
Hello,
I have a host which has 100% cpu utilization according to vCenter. But govc shows it as 50%. Can you please fix the issue?

Cheers,
       Thomas

**Comment by dougm** [comment]
Where are you looking in vCenter?  What are your vCenter/ESXi versions?

Looking at a 6.7U3 UI in Host -> Monitor -> Advanced, the total cpu% is matching up between the UI + govc host.info

**Comment by sithglan** [comment]
Hallo Doug,

> Where are you looking in vCenter?

During the last weeks, I had approx. 249 ESX Servers running 100% CPU during
production hours and boot storms. They have 40 Cores, 80 Hyperthreads and had
between 100-120 VMs with 2 vCPUs each burning CPU like hell. The summary tab of
the ESX server as well as the alerts where reporting 100% CPU.

govc reported the MHZ correctly, but the percentage incorrectly. The
maximum percentage I ever saw with govc was 50%. But the host was completly
utilized and oversubscribed by factor 2.5 or higher CPU wise. So I assume I
need to double that percentage to get the right value.

I picked one host:

Summary shows: 15.72 GHZ Capacity: 99.76 GHZ So doing the math: 15.72 / 99.76 = .157 => 15.7%

govc reports for that host:

Name:              <removed>
  Path:            <removed>
  Manufacturer:    HPE
  Logical CPUs:    80 CPUs @ 2494MHz
  Processor type:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
  CPU usage:       16871 MHz (8.5%)
  Memory:          1048220MB
  Memory usage:    634314 MB (60.5%)
  Boot time:       2020-03-04 11:11:18.943999 +0000 UTC
  State:           connected

So govc reports 8.5% but the real percentage is 15.7%.

Here is an old output during production hour, showing the same CPU model:
...
(ba) [~] grep 'CPU usage' <host.info for all hosts> | grep 50.0%
...
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
  CPU usage:       99765 MHz (50.0%)
..

> What are your vCenter/ESXi versions?

vCenter: 6.5.0
Builds:
14690228
11347054

ESXi: 6.5.0
Builds:
14320405
11925212
maybe others

Cheers,
	Thomas

**Comment by dougm** [comment]
That's alot of horsepower @sithglan !  Curious if you see the expected % if you revert this commit: https://github.com/vmware/govmomi/commit/add38bedfa24aece1d88fbf1e534541b05695258

Not sure if its version related, but its been quite alot time since I've used anything < 6.7
I'll take a closer look, but let me know if that makes a difference for you.

**Comment by sithglan** [comment]
Hello Doug,

* Doug MacEachern <notifications@github.com> [2020-04-11 22:37]:
> Curious if you see the expected % if you revert this commit:
> https://github.com/vmware/govmomi/commit/add38bedfa24aece1d88fbf1e534541b05695258

reverting the patch makes it worse (percentage gets lower):

without the revered patch:

Name:              <removed>
  Path:            <removed>
  Manufacturer:    HPE
  Logical CPUs:    80 CPUs @ 2494MHz
  Processor type:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
  CPU usage:       19573 MHz (9.8%)
  Memory:          1048220MB
  Memory usage:    634327 MB (60.5%)
  Boot time:       2020-03-04 11:11:18.943999 +0000 UTC
  State:           connected

with reverted patch:

Name:              <removed>
  Path:            <removed>
  Manufacturer:    HPE
  Logical CPUs:    160 CPUs @ 2494MHz
  Processor type:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
  CPU usage:       19573 MHz (4.9%)
  Memory:          1048220MB
  Memory usage:    634327 MB (60.5%)
  Boot time:       2020-03-04 11:11:18.943999 +0000 UTC
  State:           connected

> Not sure if its version related, but its been quite alot time since
> I've used anything < 6.7 I'll take a closer look, but let me know if
> that makes a difference for you.

Same for me, I'm running already 7.0 in my lab environment. But the
production runs with 6.5 and probably is not changing anytime soon. If
you don't make any progress, I'll hunt it down for you and provide a
patch, can't be that hard to find. I also have the luxury of having my
own lab so I can test many different ESX version within 30 minutes (the
time it takes to install vCenter).

Cheers,
        Thomas

**Comment by dougm** [comment]
ok, not likely version related, but hardware related.  The machines I'm looking at all have the same number of {Cores,Pkgs,Threads}:

```console
% govc object.collect -type h / summary.hardware.numCpu{Cores,Pkgs,Threads}
HostSystem:host-15  summary.hardware.numCpuCores    int16  8
HostSystem:host-15  summary.hardware.numCpuPkgs     int16  8
HostSystem:host-15  summary.hardware.numCpuThreads  int16  8
```

Looking at the API doc for ` HostListSummaryQuickStats`:

> overallCpuUsage - Aggregated CPU usage across all cores on the host in MHz.

Sounds like we should be doing this instead:

```diff
modified   govc/host/info.go
@@ -139,7 +139,7 @@ func (r *infoResult) Write(w io.Writer) error {
 		s := host.Summary
 		h := s.Hardware
 		z := s.QuickStats
-		ncpu := int32(h.NumCpuThreads)
+		ncpu := int32(h.NumCpuCores)
 		cpuUsage := 100 * float64(z.OverallCpuUsage) / float64(ncpu*h.CpuMhz)
 		memUsage := 100 * float64(z.OverallMemoryUsage) / float64(h.MemorySize>>20)
```

Can you give that a try and see how it looks?  Also curious to see the {Cores,Pkgs,Threads} for the machine you're looking at.

**Comment by sithglan** [comment]
Hello Doug,

> -		ncpu := int32(h.NumCpuThreads)
> +		ncpu := int32(h.NumCpuCores)

> Can you give that a try and see how it looks?

yes, that solves it:

Before:

Name:              <removed>
  Path:            <removed>
  Manufacturer:    HPE
  Logical CPUs:    80 CPUs @ 2494MHz
  Processor type:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
  CPU usage:       17598 MHz (8.8%)
  Memory:          1048220MB
  Memory usage:    634314 MB (60.5%)
  Boot time:       2020-03-04 11:11:18.943999 +0000 UTC
  State:           connected

With the above patch:

Name:              <removed>
  Path:            <removed>
  Manufacturer:    HPE
  Logical CPUs:    40 CPUs @ 2494MHz
  Processor type:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
  CPU usage:       17598 MHz (17.6%)
  Memory:          1048220MB
  Memory usage:    634314 MB (60.5%)
  Boot time:       2020-03-04 11:11:18.943999 +0000 UTC
  State:           connected

> Also curious to see the {Cores,Pkgs,Threads} for the machine you're
> looking at.

summary.hardware.numCpuCores    int16  40
summary.hardware.numCpuPkgs     int16  2
summary.hardware.numCpuThreads  int16  80

Cheers,
	Thomas

**Comment by dougm** [comment]
Thanks Thomas, PR w/ the fix: #1926
